framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,706 @@
+     * @param atf the annotated type factory of a given type system, whose
+     * type hierarchy will be used to update the field's type.
+     */
+    public static void updateInferredFieldType(
+            Node lhs, Node rhs, ClassTree classTree, AnnotatedTypeFactory atf) {
+        FieldAccessNode lhsFieldNode = (FieldAccessNode) lhs;
+        ClassSymbol classSymbol = getEnclosingClassSymbol(classTree, lhsFieldNode);
+        if (classSymbol == null) return; // Anonymous class => Ignore, for now.
+        String className = classSymbol.flatname.toString();
+
+        String jaifPath = getJaifPath(className);
+        AClass clazz = getAClass(className, jaifPath);
+
+        AField field = clazz.fields.vivify(lhsFieldNode.getFieldName());
+        AnnotatedTypeMirror lhsATM = atf.getAnnotatedType(lhs.getTree());
+        AnnotatedTypeMirror rhsATM = atf.getAnnotatedType(rhs.getTree());
+        updateAnnotationSetInScene(
+                field.type, atf, jaifPath, rhsATM, lhsATM, DefaultLocation.FIELD);
+    }
+
+    /**
+     * Updates the return type of the method methodTree in the
+     * Scene of the class with symbol classSymbol.
+     * <p>
+     * If the Scene does not contain an annotated return type for the method
+     * methodTree, then the type of the value passed to the return expression
+     * will be added to the return type of that method in the Scene.
+     * If the Scene previously contained an annotated return type for the
+     * method methodTree, its new type will be the LUB between the previous
+     * type and the type of the value passed to the return expression.
+     * <p>
+     * @param retNode the node that contains the expression returned.
+     * @param classSymbol the symbol of the class that contains the method.
+     * @param atf the annotated type factory of a given type system, whose
+     * type hierarchy will be used to update the method's return type.
+     */
+    public static void updateInferredMethodReturnType(ReturnNode retNode,
+            ClassSymbol classSymbol, MethodTree methodTree,
+            AnnotatedTypeFactory atf) {
+        if (classSymbol == null) return; // Anonymous class => Ignore, for now.
+        String className = classSymbol.flatname.toString();
+
+        String jaifPath = getJaifPath(className);
+        AClass clazz = getAClass(className, jaifPath);
+
+        AMethod method = clazz.methods.vivify(JVMNames.getJVMMethodName(methodTree));
+        // Method return type
+        AnnotatedTypeMirror lhsATM = atf.getAnnotatedType(methodTree).getReturnType();
+        // Type of the expression returned
+        AnnotatedTypeMirror rhsATM = atf.getAnnotatedType(retNode.getTree().getExpression());
+        updateAnnotationSetInScene(
+                method.returnType, atf, jaifPath, rhsATM, lhsATM,
+                DefaultLocation.RETURNS);
+    }
+
+    /**
+     * Updates the set of annotations in a location of a Scene.
```

Comment:
To understand what this method does we must first forget about AnnotatedTypeMirror and think only about Scene and .jaif files: Each location in a Scene, e.g. a method return, has a set of annotations because it can hold annotations for more than one type system. This method receives an AnnotatedTypeMirror and a Scene location and updates the set of annotations of that location in the Scene.

In the documentation I call curAnnos (an AnnotatedTypeMirror) a set of annotations. This is wrong and I will fix the documentation.